lib/Trace.js

```diff
@@ -0,0 +1,181 @@
+            this._span.getEndTime() - this._span.getStartTime() : 0;
+
+        const node = {
+            type: this._type,
+            // TODO http.uri http.path?
+            uri: Node._deriveUrl(this._span.getTags()),
+            operation: Node._deriveOperation(this._span),
+            timestamp: this._span.getStartTime() * 1000,
+            duration: duration * 1000,
+            properties: Node._tagsToProperties(this._span.getTags()),
+            nodes: this._nodes,
+        };
+
+        if (this._type === 'Consumer' || this._type === 'Producer') {
+            node.endpointType = 'HTTP';
+
+            const correlationId = this._type === 'Consumer' ? this._span.getParentId() : this._span.getId();
+            if (correlationId) {
+                node.correlationIds = [{
+                    value: correlationId,
+                    scope: 'Interaction',
+                }];
+            }
+        }
+
+        return node;
+    }
+
+    static _deriveOperation(span) {
+        const method = span.getTags()['http.method'];
+        if (method) {
+            return method;
+        }
+        return span.getOperationName();
+    }
+
+    static _deriveUrl(tags) {
+        let url = tags['http.url'];
+        if (url) {
+            return url;
+        }
+
+        url = tags['http.uri'];
+        if (url) {
+            return url;
+        }
+
+        url = tags['http.path'];
+        if (url) {
+            return url;
+        }
+
+        return null;
+    }
+
+    static _tagsToProperties(tags) {
+        const properties = [];
+        Object.keys(tags).forEach((objectKey) => {
+            properties.push({
+                name: objectKey,
+                value: tags[objectKey],
+                type: 'Text',
+            });
+        });
+        return properties;
+    }
+}
+
+class Trace {
+    constructor() {
+        this._nodes = [];
+    }
+
+    addNode(type, span, parentSpan) {
+        let nodes = this._nodes;
+
+        if (parentSpan && parentSpan.getId()) {
+            const node = this.findNode(parentSpan.getId());
+            if (node) {
+                nodes = node.getNodes();
+            }
+        }
+
+        // TODO throw exception if span is already there?
+        nodes.push(new Node(span, type));
+    }
+
+    setNodeType(type, span) {
+        const node = this.findNode(span.getId());
+        if (node) {
+            node.setType(type);
+        } else {
+            console.log('Error setNode node not found');
+        }
+    }
+
+    isParent(span) {
+        // we don't have to go deeper as the root spans are in top level array
+        for (let i = 0; i < this._nodes.length; i += 1) {
+            if (this._nodes[i].getSpan().getId() === span.getId()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    findNode(spanId) {
+        return Trace._findNode(spanId, this._nodes);
+    }
+
+    static _findNode(spanId, nodes) {
```

Comment:
I would recommend moving this function outside of the class, unless you are intentionally exposing it as part of the API. In that case, remove the `_` prefix.